rfcs/0010-eaglesong/0010-eaglesong.md

```diff
@@ -0,0 +1,335 @@
+ * [Round Function](#Round-Function)
+   * [Bit Matrix](#Bit-Matrix)
+   * [Circulant Multiplication](#Circulant-Multiplication)
+   * [Injection of Constants](#Injection-of-Constants)
+   * [Addition-Rotation-Addition](#Addition-Rotation-Addition)
+   * [Full Permutation](#Full-Permutation)
+ * [Eaglesong Hash](#Eaglesong-Hash)
+ * [Reference Implementations](#Reference-Implementations)
+
+
+<a name="Notation"></a>
+## Notation
+In the pseudocode snippets below the following notation is used:
+
+ - `//` -- denotes a comment
+ - `||` -- denotes concatenation (of bit strings)
+ - `length` -- returns the length (in bits) of the argument
+ - `%` -- modulo operator, computes the remainder of left-hand-side argument after division by the right-hand-side
+ - `zeros` -- produces a string of zero-bits of the specified length
+ - `xor` -- denotes the exclusive-or operation of equal-length bit strings
+ - `[]` -- python-style array indexing, with indexation starting (as it should) from zero; and when an upper-bound is included the range until but not including that upper bound is denoted; when not juxtaposed to an array, the expression `[a:b]` denotes the list of integers `{a, a+1, ..., b-1}`.
+
+<a name="Design-Strategies"></a>
+# Design Strategies
+
+<a name="Sponge-Construction"></a>
+## Sponge Construction
+
+Eaglesong is a [sponge construction](https://en.wikipedia.org/wiki/Sponge_function). This means that what is specified is a permutation `F` along with three variables:
+
+ - `r`, the rate : a positive integer;
+ - `c`, the capacity : another positive integer;
+ - `d`, the delimiter : a byte fixed to a nonzero value.
+
+The permutation maps `r+c` bits to `r+c` bits. After appending the delimiter byte to variable-length input to the hash function, this input is chunked into pieces `chunk[i]`, each of `r` bits along with the last one padded with zeros as necessary. The sponge construction defines a state, which consists of `r+c` bits initialized to zeros. The absorbing phase iterates over all chunks and xors the current chunk into the top `r` bits of the state before applying the permutation. In the squeezing phase, the output buffer is initialized to the empty string. Until this buffer is of the right size, the permutation is applied to the state and the top `r` bits are read out and appended to the output buffer. If the output buffer is too long, it is truncated. The next pseudocode illustrates this operation.
+
+```
+function sponge( input, output_length ):
+    // append delimiter and pad as necessary
+    input = input || d
+    while length(input) % r =/= 0 do:
+        input = input || 0
+
+    // split
+    for i in [0, length(input) / r] do:
+        chunks[i] = input[i*r : (i+1)*r]
+
+    // initialize state
+    state = zeros(r+c)
+
+    // absorb
+    for chunk in chunks do:
+        state[0:r] = state[0:r] xor chunk
+        state = F(state)
+
+    // squeeze
+    output = ""
+    while length(output) <= output_length do:
```

Comment:
Is it required that the generated output buffer must be larger than the required length?

Reply:
Actually, no -- great catch. It's perfectly fine to use `<` instead of `<=` in the while loop's conditional statement. The input-output behavior of the function is the same, but your solution achieves that with less work.

Technically, the sponge construction is defined to have an infinitely long output. Fixed output length derivated functions just truncate this infinitely long string. And of course it doesn't make much sense to run the permutation after you already have all your output bits :)